### PR TITLE
gh-94199: Remove hashlib.pbkdf2_hmac() Python implementation

### PR DESCRIPTION
**Pull request by vstinner**
Remove the pure Python implementation of hashlib.pbkdf2_hmac(),
deprecated in Python 3.10. Python 3.10 and newer requires OpenSSL
1.1.1 or newer (PEP 644), this OpenSSL version provides a C
implementation of pbkdf2_hmac() which is faster.

* Issue: gh-94199

### CODE REVIEW

**Comment by vstinner** [comment]
@tiran: I'm not sure about what I wrote: "the C implementation is safer". I don't know the rationale for removing the pure Python implementation. The docstring says that the Python implement is faster for long passwords. The deprecation message seems to say that the OpenSSL implementation is faster. So the removal is about performance, not safety?

**Comment by vstinner** [comment]
> @tiran: I'm not sure about what I wrote: "the C implementation is safer".

I replaced it with: "is **faster**" :-)

**Comment by tiran** [comment]
In this case faster is safer!

File: Lib/test/test_hashlib.py [hunk 1]
```diff
@@ -1096,15 +1096,8 @@ def _test_pbkdf2_hmac(self, pbkdf2, supported):
...
             self.assertEqual(out, self.pbkdf2_results['sha1'][0][0])
 
-    @unittest.skipIf(builtin_hashlib is None, "test requires builtin_hashlib")
-    def test_pbkdf2_hmac_py(self):
-        with warnings_helper.check_warnings():
-            self._test_pbkdf2_hmac(
-                builtin_hashlib.pbkdf2_hmac, builtin_hashes
-            )
-
     @unittest.skipUnless(hasattr(openssl_hashlib, 'pbkdf2_hmac'),
```
Comment (tiran) [inline Lib/test/test_hashlib.py]:
```suggestion
    @unittest.skipIf(openssl_hashlib is None, "requires OpenSSL bindings")
```

**Comment by tiran** [comment]
Please update documentation, too.

**Comment by vstinner** [comment]
PR rebased to fix a conflict on Doc/whatsnew/3.12.rst.

**Comment by vstinner** [comment]
@tiran: Would you mind to review the completed PR? (updated doc)

File: Doc/library/hashlib.rst [hunk 2]
```diff
@@ -300,23 +300,16 @@ include a `salt <https://en.wikipedia.org/wiki/Salt_%28cryptography%29>`_.
...
-      three times slower and doesn't release the GIL.
-
-   .. deprecated:: 3.10
+   .. versionadded:: 3.4
 
-      Slow Python implementation of *pbkdf2_hmac* is deprecated. In the
-      future the function will only be available when Python is compiled
-      with OpenSSL.
+   .. versionchanged:: 3.12
+      Remove the slow pure Python implementation, deprecated in Python 3.10.
```
Comment (tiran) [inline Doc/library/hashlib.rst]: The remark is hard to understand for users that are not aware of the history of the function. I would rephrase it to something like "pbkdf2_hmac is now only available when Python is built with OpenSSL. The slow pure Python implementation has been removed."

**Comment by rhettinger** [comment]
PEP 399 says that we should have a pure python version when possible.  That said if there is bona fide security risk, then it should be removed.

File: Doc/library/hashlib.rst (hunk 2 above)
Comment (vstinner) [reply Doc/library/hashlib.rst]: I copied your text, thanks.

**Comment by vstinner** [comment]
> PEP 399 says that we should have a pure python version when possible. That said if there is bona fide security risk, then it should be removed.

PyPy gets the C implementation of pbkdf2_hmac() from OpenSSL: https://foss.heptapod.net/pypy/pypy/-/blob/branch/py3.9/lib_pypy/_hashlib/__init__.py#L205

It has the same Python fallack implementation if the C implementation is missing: https://foss.heptapod.net/pypy/pypy/-/blob/branch/py3.9/lib-python/3/hashlib.py#L192

But I guess that the C implementaiton is always available on OpenSSL 1.1.1 and newer.

**Review by brettcannon — APPROVED** [review]
